File: mobile/test/trail_access_test.dart
import 'package:aetherride_mobile/domain/routing/trail_access.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const line = <List<double>>[
    [8.40, 49.40],
    [8.41, 49.41],
    [8.42, 49.42],
  ];

  test('nearest end when not downhill', () {
    final o = orientTrail(
      geometry: line,
      fromLat: 49.42,
      fromLng: 8.42,
      preferDownhill: false,
    );
    expect(o.reversed, isTrue);
    expect(o.usedElevation, isFalse);
    expect(o.entryLat, closeTo(49.42, 1e-9));
  });

  test('gravity uses higher endpoint as start', () {
    final o = orientTrail(
      geometry: line,
      fromLat: 49.40,
      fromLng: 8.40,
      startElevM: 200,
      endElevM: 800,
      preferDownhill: true,
    );
    expect(o.usedElevation, isTrue);
    expect(o.reversed, isTrue);
    expect(o.entryLat, closeTo(49.42, 1e-9));
    expect(o.exitLat, closeTo(49.40, 1e-9));
  });

  test('gravity keeps OSM order when first point is already the top', () {
    final o = orientTrail(
      geometry: line,
      fromLat: 49.42,
      fromLng: 8.42,
      startElevM: 900,
      endElevM: 200,
      preferDownhill: true,
    );
    expect(o.reversed, isFalse);
    expect(o.entryLat, closeTo(49.40, 1e-9));
  });

  test('tiny elev delta falls back to nearest', () {
    final o = orientTrail(
      geometry: line,
      fromLat: 49.42,
      fromLng: 8.42,
      startElevM: 410,
      endElevM: 412,
      preferDownhill: true,
    );
    expect(o.usedElevation, isFalse);
    expect(o.reversed, isTrue);
  });
}

File: src/lib/routing/trailAccess.test.ts
/**
 * Run: npx tsx src/lib/routing/trailAccess.test.ts
 */
import assert from "node:assert/strict";
import { orientTrail } from "./trailAccess";

const geo: [number, number][] = [
  [8.4, 48.0],
  [8.41, 48.01],
];

{
  const n = orientTrail({
    geometry: geo,
    fromLng: 8.4,
    fromLat: 48.0,
    preferDownhill: false,
  });
  assert.equal(n.reversed, false);
  assert.equal(n.usedElevation, false);
}

{
  const down = orientTrail({
    geometry: geo,
    fromLng: 8.4,
    fromLat: 48.0,
    startElevM: 400,
    endElevM: 900,
    preferDownhill: true,
  });
  assert.equal(down.reversed, true);
  assert.equal(down.usedElevation, true);
  assert.deepEqual(down.entry, [8.41, 48.01]);
}

{
  const alreadyTop = orientTrail({
    geometry: geo,
    fromLng: 8.4,
    fromLat: 48.0,
    startElevM: 900,
    endElevM: 400,
    preferDownhill: true,
  });
  assert.equal(alreadyTop.reversed, false);
  assert.equal(alreadyTop.usedElevation, true);
}

{
  const tiny = orientTrail({
    geometry: geo,
    fromLng: 8.4,
    fromLat: 48.0,
    startElevM: 500,
    endElevM: 504,
    preferDownhill: true,
  });
  assert.equal(tiny.usedElevation, false);
}

console.log("trailAccess.test.ts OK");
